sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py

```diff
@@ -839,6 +855,67 @@ async def update_tag_properties(
             repository=repository
         )
 
+    @distributed_trace_async
+    async def upload_blob(self, repository: str, data: IO[bytes], **kwargs) -> Tuple[str, int]:
+        """Upload an artifact blob.
+
+        :param str repository: Name of the repository.
+        :param data: The blob to upload. Note: This must be a seekable stream.
+        :type data: IO
+        :returns: The digest and size in bytes of the uploaded blob.
+        :rtype: Tuple[str, int]
+        :raises ValueError: If the parameter repository or data is None.
+        """
+        try:
+            start_upload_response_headers = cast(
+                Dict[str, str],
+                await self._client.container_registry_blob.start_upload(
+                    repository, cls=_return_response_headers, **kwargs
+                )
+            )
+            digest, location, blob_size = await self._upload_blob_chunk(
+                start_upload_response_headers['Location'], data, **kwargs
+            )
+            complete_upload_response_headers = cast(
+                Dict[str, str],
+                await self._client.container_registry_blob.complete_upload(
+                    digest=digest,
+                    next_link=location,
+                    cls=_return_response_headers,
+                    **kwargs
+                )
+            )
+        except Exception as e:
+            if repository is None or data is None:
+                raise ValueError("The parameter repository and data cannot be None.") from e
+            raise
+        return complete_upload_response_headers['Docker-Content-Digest'], blob_size
+
+    async def _upload_blob_chunk(self, location: str, data: IO[bytes], **kwargs) -> Tuple[str, str, int]:
+        hasher = hashlib.sha256()
+        blob_size = 0
+        buffer = data.read(DEFAULT_CHUNK_SIZE)
+
+        while len(buffer) > 0:
+            buffer_stream = _UnclosableBytesIO(buffer)
+            response_headers = cast(
+                Dict[str, str],
+                await self._client.container_registry_blob.upload_chunk(
+                    location,
+                    buffer_stream,
+                    cls=_return_response_headers,
+                    **kwargs
+                )
+            )
+            buffer_stream.manual_close()
```

Comment:
What if there is an exception?

Reply:
It will stop and raise the exception that we'll catch above.

Reply:
I meant do we want to catch it and close the stream?

Reply:
Yup, this needs to be in a try/finally

Reply:
> I meant do we want to catch it and close the stream?

Got you. Yes, will add.